### PR TITLE
ledger: generalize the fee field to Value

### LINKED ISSUE #1515: Be more liberal with out representation of transactions
We have some properties which we need to hold:
1. `Value`s on transaction outputs are strictly positive.
2. `Value` in fees is `Ada` only.
3. `Value` in fees is non-negative.

At the moment, we partially establish this with the types:
- In the spec, we use a non-negative quantity type for outputs and fees. We don't do this in the implementation.
- In the spec and the implementation, we use a different type for fees that is only `Ada`.

However, I think there's something to be said for liberalising this and instead enforcing these conditions in the ledger rules. Specifically:
- Outputs should just have arbitrary `Value`s on them.
- Fees should just be an arbitrary `Value`.

This is what they do in the main ledger spec, and here are a couple of related reasons that I want this.

1. System compositionality

A system with negative output values or arbitrary fee values makes *sense*, it's just not quite the system that we want. So it's nice to be able to specify the bit that's just the multi-currency part, and then have separate rules that enforce the particular conditions that we want (fees in Ada, positive outputs).

Making them orthogonal also makes them easier to reason about. For example, the rule for balancing is then exactly the same as the rule in the base ledger, which makes it easy to justify.

2. Future proofing

We already have people asking if fees will be able to be paid in currencies other than Ada. Maybe one day we will - and that change would simply amount to dropping the additional rule stating that fees must be Ada. The rest of the system would remain the same, and we wouldn't have to change the representation of anything.

Similarly for negative output values (I have an idea for where those could be used, even!).

### PR DESCRIPTION
**Pull request by michaelpj**
Fixes #1515.

I think this costs us little, and it avoids having to have a central
notion of a native currency.

Pleasingly minimal changes in the implementation - we don't really do much with fees.

### CODE REVIEW

**Comment by michaelpj** [comment]
It's still just a single value comparison. We can compute the required fee in Ada and then turn it into a Value and compare.

File: plutus-wallet-api/ledger/Ledger/Index.hs [hunk 1]
```diff
@@ -270,15 +273,22 @@ checkPositiveValues t =
     then pure ()
     else throwError $ NegativeValue t
 
+-- | Check if the fees are paid exclusively in Ada.
+checkFeeIsAda :: ValidationMonad m => Tx -> m ()
+checkFeeIsAda t =
+    if (Ada.toValue $ Ada.fromValue $ txFee t) == txFee t
+    then pure ()
+    else throwError $ NegativeValue t
```
Comment (j-mueller) [inline plutus-wallet-api/ledger/Ledger/Index.hs]: Why not `FeeNotInAda` or something? `NegativeValue` doesn't really apply here (the fee should still be positive but that's not what `checkFeeIsAda` is checking)

Comment (michaelpj) [reply plutus-wallet-api/ledger/Ledger/Index.hs]: Oops thanks, copypaste error.

**Comment by michaelpj** [comment]
Okay, looks like we are doing this in the Cardano spec, so let's do it here. Could do with a quick re-check!

**Pull request merged**